src/order/V2DutchOrder.ts

```diff
@@ -0,0 +1,454 @@
+import { getDecayedAmount } from "../utils/dutchDecay";
+
+import {
+  DutchInput,
+  DutchInputJSON,
+  DutchOutput,
+  DutchOutputJSON,
+  OrderInfo,
+  OrderResolutionOptions,
+  V2Order,
+} from "./types";
+
+export type CosignerData = {
+  decayStartTime: number;
+  decayEndTime: number;
+  exclusiveFiller: string;
+  inputOverride: BigNumber;
+  outputOverrides: BigNumber[];
+};
+
+export type CosignerDataJSON = {
+  decayStartTime: number;
+  decayEndTime: number;
+  exclusiveFiller: string;
+  inputOverride: string;
+  outputOverrides: string[];
+};
+
+export type V2DutchOrderInfo = OrderInfo & {
+  cosigner: string;
+  cosignerData: CosignerData;
+  input: DutchInput;
+  outputs: DutchOutput[];
+  cosignature: string;
+};
+
+export type V2DutchOrderInfoJSON = Omit<
+  V2DutchOrderInfo,
+  "nonce" | "input" | "outputs" | "cosignerData"
+> & {
+  nonce: string;
+  input: DutchInputJSON;
+  outputs: DutchOutputJSON[];
+  cosignerData: CosignerDataJSON;
+};
+
+type V2WitnessInfo = {
+  info: OrderInfo;
+  cosigner: string;
+  inputToken: string;
+  inputStartAmount: BigNumber;
+  inputEndAmount: BigNumber;
+  outputs: DutchOutput[];
+};
+
+const V2_DUTCH_ORDER_TYPES = {
+  V2DutchOrder: [
+    { name: "info", type: "OrderInfo" },
+    { name: "cosigner", type: "address" },
+    { name: "inputToken", type: "address" },
+    { name: "inputStartAmount", type: "uint256" },
+    { name: "inputEndAmount", type: "uint256" },
+    { name: "outputs", type: "DutchOutput[]" },
+  ],
+  OrderInfo: [
+    { name: "reactor", type: "address" },
+    { name: "swapper", type: "address" },
+    { name: "nonce", type: "uint256" },
+    { name: "deadline", type: "uint256" },
+    { name: "additionalValidationContract", type: "address" },
+    { name: "additionalValidationData", type: "bytes" },
+  ],
+  DutchOutput: [
+    { name: "token", type: "address" },
+    { name: "startAmount", type: "uint256" },
+    { name: "endAmount", type: "uint256" },
+    { name: "recipient", type: "address" },
+  ],
+};
+
+const V2_DUTCH_ORDER_ABI = [
+  "tuple(" +
+    [
+      "tuple(address,address,uint256,uint256,address,bytes)", // OrderInfo
+      "address", // cosigner
+      "tuple(address,uint256,uint256)", // input
+      "tuple(address,uint256,uint256,address)[]", // outputs
+      "tuple(uint256,uint256,address,uint256,uint256[])", // cosignerData
+      "bytes", // cosignature
+    ].join(",") +
+    ")",
+];
+
+export class V2DutchOrder extends V2Order {
+  public permit2Address: string;
+
+  constructor(
+    public readonly info: V2DutchOrderInfo,
+    public readonly chainId: number,
+    readonly _permit2Address?: string
+  ) {
+    super();
+    if (_permit2Address) {
+      this.permit2Address = _permit2Address;
+    } else if (PERMIT2_MAPPING[chainId]) {
+      this.permit2Address = PERMIT2_MAPPING[chainId];
+    } else {
+      throw new MissingConfiguration("permit2", chainId.toString());
+    }
+  }
+
+  static fromJSON(
```

Comment:
oh one thing is that the FE will need to be able to construct based on only the inner data, at least to get the permit info. Maybe need a separate object for that purpose? i.e. inner one to be used for FE pre-secondary auction order data, and outer one to be used by BE and fillers for post-secondary auction order data

can be separate PR tho as rn we just need to unblock BE stuff? though I guess URA will need that one too